### ISSUE TITLE
Testnet Node Deployment (testnet.polykey.io)

### ISSUE THREAD

**Issue by Zachaccino** [MatrixAI/Polykey#194]
### Specification

It is now time for our second attempt at testnet deployment.

![image](https://user-images.githubusercontent.com/640797/138802981-b30ccd55-3ede-4cbb-9e2b-fe644695d711.png)

We had previously already done a PK deployment on AWS using ECS back when PK was `0.0.41`. While the AWS deployment worked, we hit a lot of problems which meant we had to go through an 8 month long refactoring process over the entire codebase.

Now that the codebase is finally refactored, we're ready for the second attempt.

The AWS architecture is basically the same as before, but our configuration should be a lot more simpler. There are some changes though.

* Before we had to deal with Node root certificates, now root certificates are no longer relevant to the testnet/mainnet deployment
* We are now separating into 2 clusters of PK seed nodes: `mainnet.polykey.io` and `testnet.polykey.io`. The `mainnet` is intended for production use, and we will first prototype our testnet deployment and testnet will be where new versions of PK are tested before being released on production.
* Both mainnet and testnet seed nodes will be trusted by default, but the PK releases should default to use the mainnet and have a switch to use the testnet.
* We don't know yet whether we should be using NLB or not, we may decide not to use a NLB at all. But there shouldn't be any sort of session state that is required for P2P functionality
* NLBs cannot be used with PK clients that are debugging the testnet/mainnet nodes, because they would resolve any possible node, and in this case there is in fact network session state. Instead PK client debugging has to be done with the container IPs.
* We know that IPv6 isn't supported yet so we will have IPv4 and DNS support.
* We should be using well known ports here of `1314` UDP and `1315` TCP for the ingress port and the client port respectively.
* The PK nodes are not stateless, they do require node state. However this node state is not important to us to persist. So any EBS volume mounted into the ECS container should work. Basically we just need a mutable temporary directory. What kind of mutations are there? Well the kademlia node graph is persisted atm and is not in-memory.

### Additional context

* https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/issues/197 - old issue detailing how to configure the AWS infrastructure (we're doing it manually right now) **refer to this when starting on this issue**
* https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/issues/237 - old issue regarding old task definition envrionment variables

### Tasks

1. [x] - Upload image to ECR "elastic container registry"
   * check https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/issues/197#note_496723119
   * make sure you have the right authentication details
2. [x] - Create ECS "elastic container service" Task Definition for the new image uploaded to ECR
   * the Task Definition describes how to execute the container
   * just like how we executed in `docker run`, we will need the same parameters
   * the PK agent will need a writable directory to be the node state, if we don't specify anything this will just be a temporary scratch layer by ECS, so we should be using a volume mount of some sort, the data inside this PK agent is not important, therefore any AWS volume should be ok, however an NFS/EFS volume might help us in case we want to debug things
   * additional environment variables for unattended bootstrapping and port/host binding
     - MatrixAI/Polykey#202 - bootstrapping
     - MatrixAI/Polykey#286 - ports
3. [x] - Start the ECS service, just cluster of 1, test that it is working by using the PK CLI and directly contacting the ECS IP address and port for `PK_PORT`.
4. [x] - Integrate firewall (security group), NLB and elastic IP to the NLB and then attach the `testnet.polykey.io` domain to the NLB
   * NLB won't maintain a session between connections in order to point to the same agent, ideally we won't need to have a common for NAT-busting purposes, this shouldn't be a problem for NAT-busting for either hole-punching relay or actual relay, as far I know
   * if it is a problem, an alternative to NLB, is domain-level load balancing, where multiple EIPs are presented by `testnet.polykey.io` and they are randomised, cloudflare supports this https://www.cloudflare.com/en-au/learning/performance/what-is-dns-load-balancing/ resolve this in MatrixAI/Polykey#177 
~5. [ ] - Update reference documentation with testnet architecture including AWS, use a component diagram with relevant AWS resources~ - waiting on Pulumi... intermediate diagram available https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-1179923272

**Comment by CMCDragonkai** [comment]
@joshuakarp we need to agree on the right wording/name for this concept in both code and the URL.

* bootstrap - this confuses it with the bootstrap phase and the bootstrap command
* seed - I like this word as this makes sense with respect to other P2P apps
* broker - this word is too overloaded in the machine messaging ecosystem

If we use `seed`, we need to use that terminology everywhere so that way we don't have any confusion. That means `seed.polykey.io` for the domain name, plus in the `nodes` domain code as well as the certificate or seed node IDs and initial IP address.

We may not even need the seed certificates if we just hardcode the seed into kademlia with both the seed Node IDs and seed domains instead of IPs.

Oh and importantly we must be able to have kademlia refer to NodeAddresses through a domain name instead of an IP. A domain name can resolve to an IP or a list of IP addresses. That's going to be important for after MVP, since we our domain seed.polykey.io may point to other IPs in the future or globally distributed set of IPs.

**Comment by joshuakarp** [comment]
Agreed - `seed` makes most sense to me too, and distinguishes it from the bootstrap phase.

Can you clarify this a bit more?

> Oh and importantly we must be able to have kademlia refer to NodeAddresses through a domain name instead of an IP. A domain name can resolve to an IP or a list of IP addresses. That's going to be important for after MVP, since we our domain seed.polykey.io may point to other IPs in the future or globally distributed set of IPs.

**Comment by CMCDragonkai** [comment]
To solve this issue, we're going to have to deploy a test-net for the PK network. We can use the nixpkgs development style. The `master` branch represents the currently development PK. Whereas a `release` branch can be used staging into production.

The `master` branch can have CI/CD continuously deploy it to the test net. So whatever is the current master branch is also in released to the testnet. This does mean `master` is sort of unstable.

Additional branches can be marked for releasing to production net. I wonder if this is best done with a separate `release` or `prod` branch, or just done via tagging. The CI/CD for this should be manual though, as a manual QA should be done for this.

**Comment by CMCDragonkai** [comment]
The testnet will require its own domain:

```
seed.polykey.io - seed nodes for production releases
seed-test.polykey.io or testnet.polykey.io - testnet for master
```

This should allow us to maintain an incremental working system, and invite others to join in testing on the testnet.

**Comment by CMCDragonkai** [comment]
The `master` branch should be auto deployed to a testnet. A separate branch will be made for production deployment.

This gives a few advantages:

1. Most PRs a new devs will assume that feature requests get merged to master. Doesn't require much explanation.
2. Shows development activity front and center, and for people looking at GitHub which are devs, they will probably want to use the testnet.
3. No accidental pushes to production when it is a special branch requiring special privileges to push to it. Testing must be extensive for `master` before enabling a merge to a `prod` branch.

**Comment by CMCDragonkai** [comment]
We can reuse the terminology from blockchains and use:

```
mainnet.polykey.io
testnet.polykey.io
```

I like this as this captures the intent and intuition of P2P and blockchain devs.

**Comment by CMCDragonkai** [comment]
There's probably code changes necessary when we hit problem:

1. Environment variable setting to automate port/host settings, and also the root password and node state location during bootstrapping, should involve the `config.js`
2. We need to make use of reserved domains in our seed node list. Right now it expects that the node ids are known. We may say that we don't know the node ids, but we will trust whatever node id we get from resolving the `testnet.polykey.io` domain.
   * there are 2 things that are dynamic here: node ids and IP addresses, both of which can be resolved by trusting a reserved domain name
   * I don't like this idea, so we will instead pre-generate root certificates and root public key pairs. Then hand out the node ids and embed these into the source code through a config and `PK_SEED`. We as a company need to backup and maintain these root keys and such, and that means during the bootstrap phase, we need to be able to generate "deterministic" root keys. This is already possible by the `KeyManager`, however I believe the bootstrap phase CLI procedure needs to be updated to support this.
     - this has an advantage in that we only have to backup a set of 24/12 word phrases which we will keep secret as the keys to the testnet!!!!
3. Nodes domain will need to be able to resolve the DNS domain names, these domain name may present a set of IPs, each set of IP may be any possible NodeId. At this point in time, nodes expects a list of node id to IP address. If we want to allow IP address to be dynamic, we will need to change this to discover the mappings by querying the IP address as if it was a node, and asking what it's node id is, and checking if that node id is in one of the list of trusted node ids. This is similar procedure to our LAN "multicase discovery" MatrixAI/js-mdns#1 

@tegefaulkes

**Comment by CMCDragonkai** [comment]
Scheduled for after merge of https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/184 to master. Then we start from `master` branch of this repo!

@joshuakarp please notify here when ready!

**Comment by CMCDragonkai** [comment]
![image](https://user-images.githubusercontent.com/640797/138802981-b30ccd55-3ede-4cbb-9e2b-fe644695d711.png)

@joshuakarp please start on a component diagram here as well.

**Comment by CMCDragonkai** [comment]
Meeting with @tegefaulkes yesterday indicated that `pk agent start` would run the agent process directly in the foreground.

This would be what we would want to do inside our container image as well when launching via the task definition.

To make this deterministic and automatic (un-attended) the `pk agent start` command would either require the agent to already be bootstrapped, or it would need to be able to perform bootstrapping automatically and then start the agent.

Right now `pk bootstrap` is a specific command that bootstraps the PK node state. If `pk agent start` is called before bootstrapping is done, what is the expected behaviour? If we want to necessarily separate the 2 commands, we would have to program our TD to instead run the shell command:

```sh
pk bootstrap && pk agent start
```

Where with the node state is already bootstrapped, it would have to be ignored. However these commands are also used by end users so we have to balance UX concerns for users of the CLI itself as well.

I prefer the separation of responsibilities here. The `pk bootstrap` should be reserved for bootstrapping agent/node state. The location of this state can be anywhere, by default in human users, it's in the home directory, but in our TD, this can be provided via a `/srv` directory mounted with a persistent volume either by AWS elastic file system or EBS volume. The specific state here doesn't actually matter, our agents should be able to work with just temporary state. That is they are not stateless services since they need to mutate the disk for storing data in kademlia, but the state is not something we need to keep persisted. @joshuakarp is actually necessary for us to keep our node state in the DB? I'm wondering about the design of the kademlia and whether that can just be in-memory. I forgot if we had a discussion about this.

Whichever style we go for, we're going to need some environment variables.

In MatrixAI/Polykey#269 we already talked about the addition of `PK_SEED_NODES`.

For this, we're going to need env variables for:

* root recovery code related to MatrixAI/Polykey#202 
  - `PK_RECOVERY_CODE` and `PK_ROOT_PASSWORD`
  - also related to MatrixAI/Polykey#240 @emmacasolin can you confirm what env variables you are using?
* port bindings for grpc server, forward proxy, reverse proxy (see: https://github.com/MatrixAI/js-polykey/wiki/network#network-architecture)
  - we selected a default "preferred" port as 1314 for agent seed connections, but by default the ports are randomly selected by the OS, so 1314 is only for our usage in the seed nodes
  - old details here https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/issues/237#note_530324589
  - `PK_INGRESS_PORT` - ingress port is the reverse proxy port used for P2P
  - `PK_INGRESS_HOST` - ingress host is the reverse proxy host used for P2P
  - for NAT-busting, egress host and port may be relevant but I'm not sure, we'll know when we start MatrixAI/Polykey#159
  - the `testnet.polykey.io` would resolve to an `PK_INGRESS_HOST` as this is what is needed, however the expectation would be that the same IP would be used for both the ingress host and the client server host
  - is the grpc client server is on the same port as the agent server? Or can they be separate? Either way accessing the client does not need to go through the proxies, therefore, a specific port can be used here too:
  - `PK_SERVER_PORT` and `PK_SERVER_HOST` if there's a shared host/port being used by the client service
  - if `1314` is used for `PK_INGRESS_PORT`, then when PK CLI cannot be used to access this, as the CLI would expect direct contact with the service, instead they would require `PK_SERVER_PORT`, therefore an additional port `1315` could be used
* All env variable parsed data should be centralised into `src/config.ts`. Only there is where `process.env` should even be called.
  - beware the difference between PK CLI environment and PK agent environment, env variables set before `pk agent start` foreground process works, but other env variables set for other commands makes no difference to the agent process

Network architecture diagram reposted here:

![image](https://user-images.githubusercontent.com/640797/138816235-20b8eb6a-f64a-40eb-81ec-8137c8e17ab4.png)

@tegefaulkes @joshuakarp can you review and address my questions above with snippets pointing to the code where relevant

**Comment by emmacasolin** [comment]
>   * `PK_RECOVERY_CODE` and `PK_ROOT_PASSWORD`
>   * also related to [Automatically Retry the CLI Command upon Session Unlocking #240](https://github.com/MatrixAI/js-polykey/issues/240) @emmacasolin can you confirm what env variables you are using?

I'm currently using `PK_PASSWORD` as the place where the root password is stored, but I can change it to `PK_ROOT_PASSWORD`?

**Comment by CMCDragonkai** [comment]
If we don't have any other kind of password, perhaps `PK_PASSWORD` is better.

**Comment by joshuakarp** [comment]
> is actually necessary for us to keep our node state in the DB? I'm wondering about the design of the kademlia and whether that can just be in-memory.

Are you specifically referring to the seed node state here? If so, perhaps this doesn't need to be persisted. In-memory would be satisfactory for this case. It would mean that you can't "discover" the seed nodes as part of the kademlia process.

If you're referring to the global node state here (i.e. all known nodes from Kademlia) then this needs to be persisted. Otherwise, we completely reset our known nodes each time a keynode is restarted.

**Comment by joshuakarp** [comment]
Port speccing
---

Port bindings for grpc server, forward proxy, reverse proxy (see: https://github.com/MatrixAI/js-polykey/wiki/network#network-architecture)

1. we selected a default "preferred" port as 1314 for agent seed connections, but by default the ports are randomly selected by the OS, so 1314 is only for our usage in the seed nodes
    - This should be the **ingress** port of the reverse proxy. That is, for all seed nodes, the ingress port will always be 1314.
    - old details here https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/issues/237#note_530324589
    - `PK_INGRESS_PORT` - ingress port is the reverse proxy port used for P2P
    - `PK_INGRESS_HOST` - ingress host is the reverse proxy host used for P2P

```ts
// ReverseProxy.ts
  public async start({
    ingressHost = '0.0.0.0' as Host,
    ingressPort = 0 as Port,
    serverHost,
    serverPort,
    tlsConfig,
  }: {
    ingressHost?: Host;
    ingressPort?: Port;
    serverHost: Host;
    serverPort: Port;
    tlsConfig: TLSConfig;
  }): Promise<void> {
```

2. for NAT-busting, egress host and port may be relevant but I'm not sure, we'll know when we start MatrixAI/Polykey#159
    - the egress port of the `ForwardProxy` may be necessary here. Recall our hole-punching process (for some keynode `A` trying to connect to `B`):
      * `A`'s forward proxy sends packets to `B`'s ingress_host:ingress_port (the reverse proxy)
      * `B`'s reverse proxy sends packets to `A`'s egress_host:egress_port (the forward proxy)
   - testnet is not behind a NAT, it's on our public AWS, when the PK agent sends packets out on the egress port, the AWS firewall will automatically allow inbound packets on that egress port as long as we have sent packets out on that egress port, this is what we expect, but we should test if this is the case
3. the `testnet.polykey.io` would resolve to an `PK_INGRESS_HOST` as this is what is needed, however the expectation would be that the same IP would be used for both the ingress host and the client server host
   - by resolve here, we mean eventually get to...
   - the `PK_INGRESS_HOST` should be `0.0.0.0` because we want to bind to all interfaces within the ECS container as iti s a public service
   - we don't know what the interface address is on AWS ECS containers, so we just bind to all interfaces
4. is the grpc client server on the same port as the agent server? Or can they be separate? Either way accessing the client does not need to go through the proxies, therefore, a specific port can be used here too:
    -  Both the client server port and agent server port are currently randomised on start-up (through `GRPCServer::bindAsync()` - passing port 0 randomises the port), and are therefore separate:

```ts
  public static async createPolykey({
    // ...
    clientGrpcHost?: string;
    agentGrpcHost?: string;
    clientGrpcPort?: number;
    agentGrpcPort?: number;
    // ...
  }): Promise<Polykey> {
    const clientGrpcHost_ = clientGrpcHost ?? '127.0.0.1';
    const clientGrpcPort_ = clientGrpcPort ?? 0;
    const agentGrpcHost_ = agentGrpcHost ?? '127.0.0.1';
    const agentGrpcPort_ = agentGrpcPort ?? 0;
    // ...
  }

  public async start() {
    // ...
    // GRPC Server
    // Client server
    await this.clientGrpcServer.start({
      services: [[ClientService, this.clientService]],
      host: this.clientGrpcHost as Host,
      port: this.clientGrpcPort as Port,
      tlsConfig: {
        keyPrivatePem: keyPrivatePem,
        certChainPem: certChainPem,
      },
    });
    // Agent server
    await this.agentGrpcServer.start({
      services: [[AgentService, this.agentService]],
      host: this.agentGrpcHost as Host,
      port: this.agentGrpcPort as Port,
    }); 
    // ...
  }
```

* attempting to set the ports as the same is successful, but only if the client server is on a different host to the agent server (is this expected?):

<details>

```ts
import { PolykeyAgent } from './src';
import os from 'os';
import path from 'path';
import fs from 'fs';
import Logger, { LogLevel, StreamHandler } from '@matrixai/logger';

async function main() {
  const password = 'password';
  const logger = new Logger('PolykeyAgent Test', LogLevel.DEBUG, [
    new StreamHandler(),
  ]);
  let dataDir = await fs.promises.mkdtemp(
    path.join(os.tmpdir(), 'polykey-test-'),
  );
  const nodePath = `${dataDir}/polykey`;
  const pk = await PolykeyAgent.createPolykey({
    password,
    nodePath,
    clientGrpcHost: '127.0.0.1',
    clientGrpcPort: 1315,
    agentGrpcHost: '127.0.0.2',
    agentGrpcPort: 1315,
    logger,
    cores: 1,
    workerManager: null,
  });
  await pk.start({});

  await pk.stop();
  await pk.destroy();
  await fs.promises.rm(dataDir, {
    force: true,
    recursive: true,
  });
}

main();
```

</details>

* `PK_SERVER_PORT` and `PK_SERVER_HOST` if there's a shared host/port being used by the client service
* if `1314` is used for `PK_INGRESS_PORT`, then when PK CLI cannot be used to access this, as the CLI would expect direct contact with the service, instead they would require `PK_SERVER_PORT`, therefore an additional port `1315` could be used

**Comment by CMCDragonkai** [comment]
Example usage of `src/config.ts`:

```ts
import process from 'process';

/**
 * Polykey configuration
 */
const config = {
  version: '1.0.0', // Keep this synchronized with package.json
  stateVersion: 1,
  providers: {
    'github.com': {
      clientId: 'ca5c4c520da868387c52',
    },
  },
  ingressPort: parsePort(process.env['PK_INGRESS_PORT']),
};

function parsePort(value: string | undefined): number {
  let port = parseInt(value ?? '0');
  if (isNaN(port)) {
    port = 0;
  }
  return port;
}

export default config;
```

See how the `parsePort` function checks for both null/undefined and also if the port is `NaN`.

This function can be re-used and abstracted across other situations that require parsing a port. @emmacasolin especially on the CLI.

**Comment by CMCDragonkai** [comment]
For addressing point 1. https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-952492208

Problem regarding the usage of `src/config.ts` and `PolykeyAgent.ts`. The `PK_INGRESS_PORT` has to be propagated from env variables to the `ReverseProxy.createReverseProxy()`.

However we noticed during the integration of `js-async-init`, some things weren't done exactly the way it's supposed to be.

In `PolykeyAgent`, the structure at the end of `createPolykeyAgent` should be:

```ts
    const pk = new Polykey({ ... });
    await pk.start({ ... });
    return pk;
```

The reason to have an asynchronous `createX` static constructor is to call the `start` and ensure that constructed instances are started already. If this wasn't intended, then you only needed the `StartStop` decorator, not the `CreateDestroyStartStop` decorator.

The fact that `PolykeyAgent` wasn't done this way, may mean that other classes that are using the `CreateDestroyStartStop` pattern are also doing it incorrectly.

Now because the `createX` static methods are meant to call the asynchronous start, this means `start` parameters/options must be available in the `createX` static methods. The `ReverseProxy` has these parameters:

```ts
  public async start({
    ingressHost = '0.0.0.0' as Host,
    ingressPort = 0 as Port,
    serverHost,
    serverPort,
    tlsConfig,
  }: {
    ingressHost?: Host;
    ingressPort?: Port;
    serverHost: Host;
    serverPort: Port;
    tlsConfig: TLSConfig;
  }): Promise<void> {
```

These need to be replicated to the `createReverseProxy`.

@tegefaulkes 

Note the difference between constructor parameters and `start` parameters. In the recent work on EFS and `js-db`, it's better to put parameters into the constructor if it is expected that they will stick around as class properties. Only if the parameters are temporary to the asynchronous start, then you should have them in the `start` method. This means there are some changes required to the PK domains that haven't yet been done.

In `ReverseProxy`, there are some cases where `start` does require parameters, because for example `ingressPort` can be set to `0`, and then it has be resolved subsequently to the actual port, and that's why it is in the `start` method.

So the list of things to do here are:

1. Review all uses of `CreateDestroyStartStop` decorator and ensure that their `createX` functions are calling `start`
2. Move any parameter that is meant to be stored as a class property and doesn't require asynchronous operations on them to the constructor instead of `start`.
3. Propagate all remaining `start` parameters to the `createX` functions so that when calling create you can parameterise the asynchronous creation.
4. Make sure to test all of this in accordance to the matroska/babushka doll. Please create a sequence diagram to update while you're doing this.
5. The difference between encapsulated dependencies and just external references is whether the dependency is optional. If optional, then it is an encapsulated dependency, if required, then it is an external reference. So you don't manage it's lifecycle, which means you don't call create or start.

**Comment by CMCDragonkai** [comment]
The `src/PolykeyAgent.ts` should be `class PolykeyAgent` not `class Polykey`. The class names and file names should match.

**Comment by emmacasolin** [comment]
Would something like this be how the env variable for the root password would be stored in `src/config.ts`? There isn't really any extra checking that would need to be done here since we're already receiving a string.
```ts
import process from 'process';

/**
 * Polykey configuration
 */
const config = {
  version: '1.0.0', // Keep this synchronized with package.json
  stateVersion: 1,
  providers: {
    'github.com': {
      clientId: 'ca5c4c520da868387c52',
    },
  },
  rootPassword: process.env['PK_PASSWORD'],
};

export default config;
```

**Comment by CMCDragonkai** [comment]
Regarding point 4. in https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-952492208

GRPC servers and GRPC services are different things. You can multiple GRPC services on the same GRPC server. You can also have multiple GRPC servers running their own set of services.

Originally we had 1 GRPC server running 2 GRPC services `AgentService` and `ClientService`. We later changed this to 2 GRPC servers, one for the agent, one for the client.

The advantage of this is:

* The agent GRPC server and agent GRPC service is plain-text and only accessible over the networking proxies.
* The client GRPC server and client GRPC service is TLS-required and is accessible outside of the networking proxies.
  - this means `pk` CLI calls and GUI doesn't require going over the proxies to use the client GRPC service
  - this means it is quicker to launch and connect

In the AWS testnet architecture diagram https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-951520326, the testnet PK nodes can expose their client service for debugging purposes on the same ECS IP as used by the agent service.

We will run `pk` CLI calls from our own computers in our office or at home, and we have to be able to call client service of the testnet.

This means the GRPC server for the client service will need to be exposed on the public internet, and therefore we will have to configure them to bind to `0.0.0.0`.

In the source code with creating the `PolykeyAgent`:

```ts
  public static async createPolykey({
    // ...
    clientGrpcHost?: string;
    agentGrpcHost?: string;
    clientGrpcPort?: number;
    agentGrpcPort?: number;
    // ...
  }): Promise<Polykey> {
    const clientGrpcHost_ = clientGrpcHost ?? '127.0.0.1';
    const clientGrpcPort_ = clientGrpcPort ?? 0;
    const agentGrpcHost_ = agentGrpcHost ?? '127.0.0.1';
    const agentGrpcPort_ = agentGrpcPort ?? 0;
    // ...
  }
```

The `clientGrpcHost` and `agentGrpcHost` is currently defaulted to `127.0.0.1`. This is correct for general-usecase. But for our testnet usage, we must configure the `clientGrpcHost` to `0.0.0.0` in order to achieve the above.

Therefore we need an environment variable configuration for this as well.

* `PK_CLIENT_HOST`
* `PK_CLIENT_PORT`

```ts
  ingressHost: process.env['PK_INGRESS_HOST'] ?? '0.0.0.0',
  ingressPort: parsePort(process.env['PK_INGRESS_PORT']),
  clientHost: process.env['PK_CLIENT_HOST'] ?? '127.0.0.1',
  clientPort: parsePort(process.env['PK_CLIENT_PORT']),
```

**Comment by CMCDragonkai** [comment]
Regarding the usage of `src/config.ts`. The reading from environment variables is a side-effect. This means `import config from './config.ts'` is already performing side effects.

Ideally this performance of side effects is done in the `src/bin` scripts. Therefore, instead of importing config, all relevant configuration parameters should be dependency injected into `PolykeyAgent` and `PolykeyClient`.

This also simplifies testing, as this means we do not need to mock the process environment.

Atm, the only usages are:

```
[nix-shell:~/Projects/js-polykey/src]$ ag 'import config'
keys/utils.ts
30:import config from '../config';

PolykeyAgent.ts
30:import config from './config';
```

So these 2 must be removed, the `keys` function `generateCertificate` should take a `version` parameter instead of importing the `config`.

The relevant bin CLI commands that will be `import config` should be `pk agent start`.

All parameters in `config` should be DIed. Don't do DI the whole `config` object, be more specific to the relevant parameters. The `src/bin` scripts can remap `config` properties to whatever DI parameters are required.

The only place that does `PolykeyAgent.createPolykeyAgent` should be `src/bin/agent/start.ts`. Every other command is doing `PolykeyClient.createPolykeyClient`. Does the client need configuration parameters? Not sure, but if they do, then the relevant scripts should be called to do so.

Note that `pk bootstrap` does not create client or agent, but it has a bunch of comments that should be cleaned up. So the code is made clearer what it's doing. The `PK_PASSWORD` and `PK_RECOVERY_CODE` do not need to be in `src/config.ts` since they don't configure anything across the application.

---

If `src/bin/agent/start.ts` is the ONLY place that imports `src/config.ts`, then let's simplify this, and just copy that all into `src/bin/agent/start.ts`. There's no need to have a `src/config.ts` file AT ALL!

**Comment by CMCDragonkai** [comment]
@joshuakarp can you update the issue description here and collate the above comments into a consistent spec.

**Comment by CMCDragonkai** [comment]
We can keep `src/config.ts` for ONLY static configuration, and if so a block comment should be used there to tell devs to not put `process.env` processing there. So that `process.env` related side-effects are centralised into `src/bin`.

**Comment by joshuakarp** [comment]
Incorporated these points https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-952530723 regarding fixing the `CreateDestroyStartStop` decorator into the description. 

However, it would most likely be best to move this to its own issue (blocking this issue), but separate from this testnet deployment.

**Comment by CMCDragonkai** [comment]
@joshuakarp we need to incorporate my comments regarding the usage of `config.ts` as well. Keep it as a static, and we should have env variables being process/parsed in the `src/bin`. Don't forget to incorporate the diagrams as well.

**Comment by joshuakarp** [comment]
Yeah, was still in progress with that. Added tasks regarding `config.ts` from https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-952547227, as well as some clarifications on the ports. Still working on component diagram.

**Comment by joshuakarp** [comment]
Removed all top-level issue description tasks related to async-init fixes into MatrixAI/Polykey#281.

**Comment by CMCDragonkai** [comment]
Based on this graph plan I'm going to structure this epic and apply estimates onto the roadmap: ![feh_158766_000001_IMG_20211101_094401](https://user-images.githubusercontent.com/640797/139618977-0f082911-d871-47e4-a7d0-74f6f30b641b.jpg)

Some notes:

1. @tegefaulkes on block 1 MatrixAI/Polykey#202 and block 2 MatrixAI/Polykey#286
2. @joshuakarp on block 3 MatrixAI/Polykey#269
3. @emmacasolin and @scottmmorris on supporting work this time, the rest of the AWS blocks are in this issue and will be continued by @tegefaulkes and @joshuakarp
4. block 1, 2, 3 can branch off master for now, and will need to rebase on top when MatrixAI/Polykey#275 and  `command-retry-loop` is merged
5. Over time move towards the pattern of nested encapsulated dependencies, vs required dependencies in 2 diagrams: component and sequence.

**Comment by CMCDragonkai** [comment]
Procedural tasks in this epic are blocked by MatrixAI/Polykey#231 and MatrixAI/Polykey#285. All the AWS architecture work is being recorded here.

**Comment by CMCDragonkai** [comment]
Need a separate issue regarding the port binding environment variables for `pk agent start`, this will be required to actually ensure we are binding correctly. This corresponds to block 2 in the diagram. Separating some things from MatrixAI/Polykey#202. This is now in MatrixAI/Polykey#286.

**Comment by CMCDragonkai** [comment]
The `config.ts` task isn't relevant to this issue. I'm incorporating that in the merging of the last MR: https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/213. So I'll move that task out.

**Comment by CMCDragonkai** [comment]
The usage of an `agentPB` and `clientPB` file is made problematic by versioning. So I'm removing this and making code use `utilsPB`, `vaultsPB`... etc. This makes it possible in the future for us to create `index.ts` under `src/proto` which could export potential versions of particular PB files. However I'd want to make it possible for us not have to import everything as the application gets larger, so separating by files is useful.

```ts
import { clientPB, utilsPB } from '../proto/v1';
```

Anyway I'm leaving this to be figured out later. For now as long as the code is using `utilsPB` or `vaultsPB` this is suffiicent as in the future, we make changes while preserving this usage.

I need to find a quick way to refactor the entire codebase when doing these changes.

**Comment by CMCDragonkai** [comment]
The `package.json` `version` is now synchronised to `src/config.ts` too https://github.com/MatrixAI/TypeScript-Demo-Lib/pull/33. This is being merged in https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/213.

**Comment by joshuakarp** [comment]
Start date changed from Nov 10th to Nov 15th (based on delays in MatrixAI/Polykey#231).

**Comment by joshuakarp** [comment]
Start date changed from Tuesday Nov 16th to Thursday Nov 18th (delays in MatrixAI/Polykey#269, MatrixAI/Polykey#231, and CLI MR on Gitlab). Also adjusted end date to cater for weekend (from Sunday Nov 21st to Tuesday Nov 23rd).

**Comment by joshuakarp** [comment]
Start date changed from Thursday Nov 18th to Wednesday Dec 1 (delayed from refactoring work in MatrixAI/Polykey#283).

**Comment by CMCDragonkai** [comment]
This is scheduled to be done as soon as MatrixAI/Polykey#310 is merged into master. Then the `src/config.ts` will have properly maintained `testnet` nodes.

**Comment by joshuakarp** [comment]
The PR that addresses this issue should also split the nodes agent service tests(`nodesChainDataGet` and `nodesClosestLocalNode`) as per https://github.com/MatrixAI/js-polykey/pull/310#issuecomment-1032003030.

**Comment by CMCDragonkai** [comment]
Not split tests, no tests have been written there, so this PR should introduce new tests.

> The only nodes tests that I haven't split yet are the agent service tests for nodesChainDataGet, nodesClosestLocalNode, and nodesHolePunchMessage because there are no tests for those written yet.

**Comment by CMCDragonkai** [comment]
Make sure that MatrixAI/Polykey#326 has the task list updated.

**Comment by joshuakarp** [comment]
> Not split tests, no tests have been written there, so this PR should introduce new tests.
> 
> > The only nodes tests that I haven't split yet are the agent service tests for nodesChainDataGet, nodesClosestLocalNode, and nodesHolePunchMessage because there are no tests for those written yet.

My mistake - I misinterpreted this as only `nodesHolePunchMessage` not having tests yet. Will update.

**Comment by joshuakarp** [comment]
The latest pipeline for master (6 days ago when MatrixAI/Polykey#321 was merged): https://gitlab.com/MatrixAI/open-source/js-polykey/-/commit/a54598eca29a9b89d1d169e2fd9326c469139dc4

It is a working build, but some test failures are apparent (presumably will be fixed once MatrixAI/Polykey#310 and MatrixAI/Polykey#320 are merged into master).

**Comment by CMCDragonkai** [comment]
We only nodes tests to be passing to proceed. The vaults and discovery stuff won't block testnet deployment. So you should continue reviewing the docs. You should start on AWS (you have the login?) once MatrixAI/Polykey#310 is merged.

**Comment by joshuakarp** [comment]
Our most recent pipeline now that MatrixAI/Polykey#310 and MatrixAI/Polykey#320 have been merged: https://gitlab.com/MatrixAI/open-source/js-polykey/-/pipelines/470214101

Only a few test failures that will be handled in external PRs. So can push ahead with this.

**Comment by joshuakarp** [comment]
I'm going to start by simply attempting to deploy TypeScript-Demo-Lib on AWS and playing around with it. Will spin it down once it's finished.

**Comment by CMCDragonkai** [comment]
> the PK agent will need a writable directory to be the node state, if we don't specify anything this will just be a temporary scratch layer by ECS, so we should be using a volume mount of some sort, the data inside this PK agent is not important, therefore any AWS volume should be ok, however an NFS/EFS volume might help us in case we want to debug things

Previous TD had a EFS attached here. However some things are different now:

* We have recovery code which allows us to deterministically regenerate the keypair. This is essential when redeploying new versions of the seed nodes, we need to keep this.
* The internal state of seed nodes is not super important. Things like the GG, ACL... etc should all pretty much remain pristine. The only thing the seed nodes are doing is mutating its node graph.
* The node graph is more cache than anything. It doesn't necessarily need to be preserved.
* Because caching the state can be useful for performance, instead of relying on the scratch layer which writes to memory, we can add a EBS volume mount rather than NFS/EFS.

@joshuakarp you should make the modification to the task definition now that we have successfully tested contacting the remote agent.

**Comment by CMCDragonkai** [comment]
Testnet fully deployed... but ongoing tests are needed.

**Comment by CMCDragonkai** [comment]
Initial AWS architecture diagram:

![initial aws arch](https://user-images.githubusercontent.com/640797/178182690-b6969f2b-bfd8-470f-b0a8-ae4f03413f7b.svg)

This is not fully comprehensive simply because the cloudcraft doesn't sufficiently capture all the VPC and subnet architecture and the association with 3 NLBs and availability zones.

We won't do that until we have moved to using Pulumi, and then we can get a better auto generated architecture. For now the intermediate diagram is sufficient for us to close this issue.

**Comment by CMCDragonkai** [comment]
A summary...

Relay/Seed/Bootstrap/Testnet Deployment (it's taken at least 4 attempts):

* 2021 February - March: The First Attempt for `relay.polykey.io`: Ning and Robert and Gideon: Using AWS and ECS with certificates
* 2021 July - November: - The Second Attempt for `bootstrap/seed.polykey.io` after Major Refactoring: New problems involving unattended deterministic bootstrapping with recovery key, CI/CD building the docker image, trusted seed nodes
* 2022 February - The Third Attempt for `testnet.polykey.io`: Here we have a successful connection from client to agent running on AWS through a manual deployment
* 2022 July - The Fourth Attempt for `testnet.polykey.io`: Automatic deployment now with NLBs, CI/CD and Polykey-Infrastructure, new problem is fixing the P2P network

**Comment by CMCDragonkai** [comment]
* October 2022 - The Fifth Attempt using Polykey-Infrastructure but switched from ECS instances to EC2 instances with fixed pool of recovery codes for deterministic seed node keypairs. Removed NLBs. 20 minute bootstrapping time went to 20 seconds bootstrapping time, then with crypto upgrade less than 5 seconds. New problems:
  - GRPC needs to be removed in favour of our own RPC solution (transport agnostic RPC)
  - uTP must be removed in favour of QUIC (data transfer layer)
  - Nodes and connection management needs to be bidirectional
  - Local discovery needs MDNS
  - Local traversal needs multi-address with local address prioritisation, otherwise we hit things like NAT hair pinning
